### PR TITLE
[flutter_tools] add support for faster incremental build

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Introduce BuildSystem.buildIncremental, which keeps an in-memory cache of timestamps instead of file hashes. This will be used to implement localization generation through the same build system API.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
FYI @zanderso as a solution for https://github.com/flutter/flutter/pull/55241

**Comment by zanderso** [comment]
The `buildIncremental` addtion to the `BuildSystem` API seems like an elegant solution to the problem.

File: packages/flutter_tools/test/general.shard/build_system/file_store_test.dart [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import 'package:flutter_tools/src/artifacts.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/logger.dart';
+import 'package:flutter_tools/src/build_system/file_store.dart';
+import 'package:mockito/mockito.dart';
+
+import '../../src/common.dart';
+
+void main() {
+  test('FileStore initializes file cache', () {
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/build_system/file_store_test.dart:18]: Can these be `testWithoutContext`?

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 2]
```diff
@@ -939,9 +988,9 @@ class Node {
       }
 
       final String absolutePath = file.path;
-      final String previousHash = fileHashStore.previousHashes[absolutePath];
-      if (fileHashStore.currentHashes.containsKey(absolutePath)) {
-        final String currentHash = fileHashStore.currentHashes[absolutePath];
+      final String previousHash = fileHashStore.previousMarks[absolutePath];
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/build_system.dart:991]: wasn't sure what to call it. Essentially just a string key checked for `==`. Maybe just `assetKey` instead.

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 3]
```diff
@@ -569,6 +569,55 @@ class BuildSystem {
...
+  /// Perform an incremental build of `target` and all of its dependencies.
+  ///
+  /// If [previousBuild] is not provided, a new incremental build is
+  /// initialized.
+  Future<BuildResult> buildIncremental(
+    Target target,
+    Environment environment,
+    BuildResult previousBuild,
+  ) async {
+      environment.buildDir.createSync(recursive: true);
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/build_system.dart]: Fixed

File: packages/flutter_tools/lib/src/build_system/build_system.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/build_system.dart]: renamed to assetKey

File: packages/flutter_tools/lib/src/build_system/file_store.dart [hunk 4]
```diff
@@ -63,49 +63,63 @@ class FileHash {
...
+
+/// A globally accessible cache of files.
 ///
 /// In cases where multiple targets read the same source files as inputs, we
 /// avoid recomputing or storing multiple copies of hashes by delegating
-/// through this class. All file hashes are held in memory during a build
-/// operation, and persisted to cache in the root build directory.
+/// through this class.
+///
+/// This class will use either timestamps or file hashes depending on the
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/file_store.dart]: Done

File: packages/flutter_tools/test/general.shard/build_system/file_store_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/build_system/file_store_test.dart]: Done

**Review by zanderso — APPROVED** [review]
lgtm

**Comment by jonahwilliams** [comment]
Both pending tests passed but did not phone home, landing

**Pull request merged**